src/Adapter/Order/OrderProductUpdater.php

```diff
@@ -0,0 +1,545 @@
+                        || (
+                            $product->pack_stock_type == Pack::STOCK_TYPE_DEFAULT
+                            && (Configuration::get('PS_PACK_STOCK_TYPE') == Pack::STOCK_TYPE_PACK_ONLY
+                                || Configuration::get('PS_PACK_STOCK_TYPE') == Pack::STOCK_TYPE_PACK_BOTH)
+                        )
+                    ) {
+                        $manager->addProduct(
+                            $orderDetail->product_id,
+                            $orderDetail->product_attribute_id,
+                            new Warehouse($movement['id_warehouse']),
+                            $quantityToReinject,
+                            null,
+                            $movement['price_te']
+                        );
+                    }
+                } else {
+                    $manager->addProduct(
+                        $orderDetail->product_id,
+                        $orderDetail->product_attribute_id,
+                        new Warehouse($movement['id_warehouse']),
+                        $quantityToReinject,
+                        null,
+                        $movement['price_te']
+                    );
+                }
+            }
+
+            $productId = $orderDetail->product_id;
+
+            if ($delete) {
+                $orderDetail->delete();
+            }
+
+            StockAvailable::synchronize($productId);
+        } elseif ($orderDetail->id_warehouse == 0) {
+            StockAvailable::updateQuantity(
+                $orderDetail->product_id,
+                $orderDetail->product_attribute_id,
+                $quantityToReinject,
+                $orderDetail->id_shop,
+                true,
+                [
+                    'id_order' => $orderDetail->id_order,
+                    'id_stock_mvt_reason' => Configuration::get('PS_STOCK_CUSTOMER_RETURN_REASON'),
+                ]
+            );
+
+            // sync all stock
+            (new StockManager())->updatePhysicalProductQuantity(
+                (int) $orderDetail->id_shop,
+                (int) Configuration::get('PS_OS_ERROR'),
+                (int) Configuration::get('PS_OS_CANCELED'),
+                null,
+                (int) $orderDetail->id_order
+            );
+
+            if ($delete) {
+                $orderDetail->delete();
+            }
+        } else {
+            throw new OrderException('This product cannot be re-stocked.');
+        }
+    }
+
+    /**
+     * @param Order $order
+     * @param OrderDetail $orderDetail
+     * @param int $oldQuantity
+     *
+     * @throws OrderException
+     */
+    private function updateCustomizationOnProductDelete(Order $order, OrderDetail $orderDetail, int $oldQuantity): void
+    {
+        if (!(int) $order->getCurrentState()) {
+            throw new OrderException('Could not get a valid Order state before deletion');
+        }
+
+        if ($order->hasBeenDelivered()) {
```

Comment:
Are you this use case should happen? I thought updates were blocked on delievered/paid orders? 🤔

This kinf of updates only happen when refunding products but it's another feature/handler

Reply:
So after a quick check in the handler L216, udpates are blocked for **delivered** orders, same for deleting products L279, but I guess it can happen for a paid order not yet delivered So only one use case to manage here
Unless this product updater could be used in another handler?

Reply:
done